### PR TITLE
add stream log to java DSL

### LINKED ISSUE #4275: Add stream log and stream logs by app name support to java DSL
**Problem description:**
According  to [API docs](https://docs.spring.io/spring-cloud-dataflow/docs/current/reference/htmlsingle/#api-guide-resources-stream-logs-by-app-name) there's an API to get the stream log and log for a specific app within a stream.

This is  missing from the java DSL 

**Solution description:**
Add endpoints for java DSL

**Description of alternatives:**
N/A

**Additional context:**
N/A

### PR DESCRIPTION
**Pull request by DemianTinkiel**
Fixes #4275

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@DemianTinkiel Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-dataflow&pullRequestId=4276)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-dataflow&pullRequestId=4276) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by sabbyanandan** [comment]
Nice work, @DemianTinkiel! We will have a look shortly.

**Comment by pivotal-issuemaster** [comment]
@DemianTinkiel Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-dataflow&pullRequestId=4276)!

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamTemplate.java [hunk 1]
```diff
@@ -155,6 +162,16 @@ public void scaleApplicationInstances(String streamName,
 		restTemplate.postForObject(url, properties, Object.class);
 	}
 
+	@Override
+	public String streamExecutionLog(String streamName) {
+		return restTemplate.getForObject(logsLink.expand(streamName, null).getHref(), String.class);
```
Comment (ilayaperumalg) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamTemplate.java:167]: Have you tried this with an existing stream with applications deployed? I wonder if we can use the explicit `streams/logs/{streamName}` endpoint for this.

**Comment by ilayaperumalg** [comment]
LGTM

**Comment by DemianTinkiel** [comment]
Hi @ilayaperumalg @sabbyanandan  @tzolov , is there anything else I should do to get the PR merged in? Also if you don't mind me asking, when can I expect it released? I understand you have release cycles to adhere to, I just want to know when to start depending on it the "official way" instead of my own copy. Thanks!

**Comment by tzolov** [comment]
Hi @DemianTinkiel, thanks for the very useful contribution! 

I've been reviewing but got puzzled by the: `Stream#logs(StreamApplication app)` syntax and its corresponding `streams/logs/{streamName}/{appName}` REST API. 

From what i can see the `{appName}` is not a `StreamApplication` but looks more like a runtime app name, (e.g. the runtime app names returned by the `/runtime/streams`).

IMO,  this is a problem of the `streams/logs/{streamName}/{appName}` API   <<< @ilayaperumalg 

Still until the ambiguity in the REST API is streamlined we should use a parameter like `String runtimeAppName` instead of `StreamApplication app)`. What do you think?

**Comment by tzolov** [comment]
hold on @DemianTinkiel , 
i think we might be able to fix this within the Stream.getLogs(app) implementation itself. I believe we can retrieve the app's `deploymentId` (used by the REST API) form the StreamApplication name. I will give it a try.  
If it works the DSL will provide a proper abstraction,  despite the confusing rest api underneath.

**Comment by DemianTinkiel** [comment]
Hi @tzolov, thanks for the review.  I don't mind making the changes to the `StreamTemplate ` 
but are you sure about
> Still until the ambiguity in the REST API is streamlined we should use a parameter like `String runtimeAppName` instead of `StreamApplication app)`. What do you think?

?
That would make `Stream#logs(StreamApplication)` the odd duck in relation to the other fluent APIs.

I see that there is `Stream#runtimeApps` (seems to only be used for testing?)`. Perhaps we can do something with that. Let me know what you prefer

**Comment by tzolov** [comment]
> That would make Stream#logs(StreamApplication) the odd duck in relation to the other fluent APIs.
Indeed but this oddness comes from the existing `streams/logs/{streamName}/{appName}` implementation. The `{appName}` is in fact a `deploymentId`.  
The proper fix would be for the REST API impl to retrieve the `deploymentId` internally using the stream and app names. 
But as this will be a breaking API change i don't want to rush it. 

In the meantime (as i've suggested above) we can do the appName to deploymentId conversion inside the DSL logs implementation instead. So the user can do `stream.logs(new StreamApplication("transformer")`.  
I am working on this right now and will share the update shortly. 

The `Stream#runtimeApps` is unfortunately a yet another hack that surface the need of a `AppRuntime` abstraction in the DSL.

**Comment by tzolov** [comment]
@DemianTinkiel check here the workaround i've added on top of your PR: https://github.com/tzolov/spring-cloud-dataflow/commit/40906bb4ad4d1218daaa2be58a51391d705c7d85

If you are happy with those changes i am ready to merge the PR.

**Comment by DemianTinkiel** [comment]
@tzolov   looks good, thank you. :+1:

**Comment by tzolov** [comment]
Updated, rebased and merged at 55936f24a280b14406370154aa87a0faca896149

@DemianTinkiel thank you for the contribution!

**Pull request closed without merging**